File: android/app/src/main/kotlin/me/dsdogs/turna/anki/reviewer/OfficialAnkiMime.kt
package me.dsdogs.turna.anki.reviewer

object OfficialAnkiMime {
    fun mimeForName(name: String): String {
        val lower = name.lowercase()
        return when {
            lower.endsWith(".png") -> "image/png"
            lower.endsWith(".jpg") || lower.endsWith(".jpeg") -> "image/jpeg"
            lower.endsWith(".gif") -> "image/gif"
            lower.endsWith(".webp") -> "image/webp"
            lower.endsWith(".svg") -> "image/svg+xml"
            lower.endsWith(".mp3") -> "audio/mpeg"
            lower.endsWith(".ogg") -> "audio/ogg"
            lower.endsWith(".wav") -> "audio/wav"
            lower.endsWith(".m4a") -> "audio/mp4"
            lower.endsWith(".mp4") -> "video/mp4"
            lower.endsWith(".webm") -> "video/webm"
            lower.endsWith(".css") -> "text/css"
            lower.endsWith(".js") -> "text/javascript"
            lower.endsWith(".html") -> "text/html"
            lower.endsWith(".json") -> "application/json"
            lower.endsWith(".woff") -> "font/woff"
            lower.endsWith(".woff2") -> "font/woff2"
            lower.endsWith(".ttf") -> "font/ttf"
            lower.endsWith(".otf") -> "font/otf"
            else -> "application/octet-stream"
        }
    }

    fun encodingForMime(mime: String): String? {
        val base = mime.substringBefore(';').trim().lowercase()
        return if (base.startsWith("text/") ||
            base == "application/json" ||
            base == "application/javascript"
        ) {
            "utf-8"
        } else {
            null
        }
    }
}

File: lib/application/anki_official/render/official_anki_mime.dart
class OfficialAnkiMime {
  OfficialAnkiMime._();

  static const binaryTypes = <String>{
    'image/png',
    'image/jpeg',
    'image/gif',
    'image/webp',
    'image/svg+xml',
    'audio/mpeg',
    'audio/ogg',
    'audio/wav',
    'audio/mp4',
    'video/mp4',
    'video/webm',
    'font/woff',
    'font/woff2',
    'font/ttf',
    'font/otf',
    'application/font-woff',
    'application/octet-stream',
  };

  static String mimeForName(String name) {
    final lower = name.toLowerCase();
    if (lower.endsWith('.png')) return 'image/png';
    if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg';
    if (lower.endsWith('.gif')) return 'image/gif';
    if (lower.endsWith('.webp')) return 'image/webp';
    if (lower.endsWith('.svg')) return 'image/svg+xml';
    if (lower.endsWith('.mp3')) return 'audio/mpeg';
    if (lower.endsWith('.ogg')) return 'audio/ogg';
    if (lower.endsWith('.wav')) return 'audio/wav';
    if (lower.endsWith('.m4a')) return 'audio/mp4';
    if (lower.endsWith('.mp4')) return 'video/mp4';
    if (lower.endsWith('.webm')) return 'video/webm';
    if (lower.endsWith('.css')) return 'text/css';
    if (lower.endsWith('.js')) return 'text/javascript';
    if (lower.endsWith('.html')) return 'text/html';
    if (lower.endsWith('.json')) return 'application/json';
    if (lower.endsWith('.woff')) return 'font/woff';
    if (lower.endsWith('.woff2')) return 'font/woff2';
    if (lower.endsWith('.ttf')) return 'font/ttf';
    if (lower.endsWith('.otf')) return 'font/otf';
    return 'application/octet-stream';
  }

  /// Image/audio/video/font use null (binary). Text uses utf-8.
  static String? encodingForMime(String mime) {
    final base = mime.split(';').first.trim().toLowerCase();
    if (base.startsWith('text/') ||
        base == 'application/json' ||
        base == 'application/javascript') {
      return 'utf-8';
    }
    return null;
  }

  static bool isUnknown(String mime) => mime == 'application/octet-stream';
}
